utils/jobstats/jobstats.py

```diff
@@ -0,0 +1,225 @@
+                "dur": self.dur_usec,
+                "args": self.jobargs}
+
+    def start_timestr(self):
+        """Return a formatted timestamp of the job's start-time"""
+        t = datetime.datetime.fromtimestamp(self.start_usec / 1000000.0)
+        return t.strftime("%Y-%m-%d %H:%M:%S")
+
+    def end_timestr(self):
+        """Return a formatted timestamp of the job's end-time"""
+        t = datetime.datetime.fromtimestamp((self.start_usec +
+                                             self.dur_usec) / 1000000.0)
+        return t.strftime("%Y-%m-%d %H:%M:%S")
+
+    def pick_lnt_metric_suffix(self, metric_name):
+        """Guess an appropriate LNT metric type for a given metric name"""
+        if "BytesOutput" in metric_name:
+            return "code_size"
+        if "RSS" in metric_name or "BytesAllocated" in metric_name:
+            return "mem"
+        return "compile"
+
+    def to_lnt_test_obj(self, args):
+        """Return a JSON-formattable object fitting LNT's 'submit' format"""
+        run_info = {
+            "run_order": str(args.lnt_order),
+            "tag": str(args.lnt_tag),
+        }
+        run_info.update(dict(args.lnt_run_info))
+        stats = self.stats
+        return {
+            "Machine":
+            {
+                "Name": args.lnt_machine,
+                "Info": dict(args.lnt_machine_info)
+            },
+            "Run":
+            {
+                "Start Time": self.start_timestr(),
+                "End Time": self.end_timestr(),
+                "Info": run_info
+            },
+            "Tests":
+            [
+                {
+                    "Data": [v],
+                    "Info": {},
+                    "Name": "%s.%s.%s.%s" % (args.lnt_tag, self.module,
+                                             k, self.pick_lnt_metric_suffix(k))
+                }
+                for (k, v) in stats.items()
+            ]
+        }
+
+
+def load_stats_dir(path, select_module=[]):
+    """Loads all stats-files found in path into a list of JobStats objects"""
+    jobstats = []
+    auxpat = (r"(?P<module>[^-]+)-(?P<input>[^-]+)-(?P<triple>[^-]+)" +
+              r"-(?P<out>[^-]+)-(?P<opt>[^-]+)")
+    fpat = (r"^stats-(?P<start>\d+)-swift-(?P<kind>\w+)-" +
+            auxpat +
+            r"-(?P<pid>\d+)(-.*)?.json$")
+    for root, dirs, files in os.walk(path):
+        for f in files:
+            m = re.match(fpat, f)
+            if m:
+                # NB: "pid" in fpat is a random number, not unix pid.
+                mg = m.groupdict()
+                jobkind = mg['kind']
+                jobid = int(mg['pid'])
+                start_usec = int(mg['start'])
+                module = mg["module"]
+                if len(select_module) != 0 and module not in select_module:
+                    continue
+                jobargs = [mg["input"], mg["triple"], mg["out"], mg["opt"]]
+
+                j = json.load(open(os.path.join(root, f)))
```

Comment:
Hmmm... shouldn't this be in a with statement? I have never seen someone do this as such.

Reply:
sure